clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java

```diff
@@ -2480,4 +2508,38 @@ String getClientId() {
     boolean updateAssignmentMetadataIfNeeded(final Timer timer) {
         return updateAssignmentMetadataIfNeeded(timer, true);
     }
+
+    /**
+     * Converts the fetched metadata to the public API ConsumerRecords.Metadata.
+     * Also computes overrides from left to the right in the case of multiple args.
+     * Any of the arg maps may be null, in which case, we just skip it.
+     */
+    @SafeVarargs
+    private static Map<TopicPartition, ConsumerRecords.Metadata> extractMetadata(final Map<TopicPartition, FetchedRecords.FetchMetadata>... nullableMetadataMaps) {
+        Map<TopicPartition, ConsumerRecords.Metadata> result = null;
+        for (Map<TopicPartition, FetchedRecords.FetchMetadata> nullableMetadataMap : nullableMetadataMaps) {
+            if (nullableMetadataMap == null || nullableMetadataMap.isEmpty()) {
+                continue;
+            } else {
+                if (result == null) {
+                    result = new HashMap<>(nullableMetadataMap.size());
+                }
+                for (Map.Entry<TopicPartition, FetchedRecords.FetchMetadata> entry : nullableMetadataMap.entrySet()) {
+                    result.put(
+                        entry.getKey(),
+                        new ConsumerRecords.Metadata(
```

Comment:
Not sure why we need to have `FetchedRecords.FetchMetadata`? It is always converted to `ConsumerRecords.Metadata` directly. The `records` in `FetchedRecords` is `ConsumerRecord`. Maybe we can make `FetchedRecords` use `ConsumerRecords.Metadata` also.